cdk/typescript/cloudfront-functions/lib/demo-cloudfront-functions-stack.ts

```diff
@@ -0,0 +1,87 @@
+import * as cdk from 'aws-cdk-lib';
+import {Construct} from 'constructs';
+import {BlockPublicAccess, Bucket, BucketEncryption} from "aws-cdk-lib/aws-s3";
+import {BucketDeployment, Source} from "aws-cdk-lib/aws-s3-deployment";
+import {Distribution, Function, FunctionCode, FunctionRuntime, OriginAccessIdentity} from "aws-cdk-lib/aws-cloudfront";
+import {CanonicalUserPrincipal, PolicyStatement} from "aws-cdk-lib/aws-iam";
+import {BehaviorOptions} from "aws-cdk-lib/aws-cloudfront/lib/distribution";
+
+export class DemoCloudfrontFunctionsStack extends cdk.Stack {
+  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
+    super(scope, id, props);
+
+    // create a bucket to deploy the website files
+    const bucket = new Bucket(this, 'WebsiteBucket', {
+      encryption: BucketEncryption.S3_MANAGED,
+      versioned: true,
+      enforceSSL: true,
+      blockPublicAccess: BlockPublicAccess.BLOCK_ALL
+    });
+
+    // create a s3 bucket deployment to deploy the website directory's files to the website bucket
+    new BucketDeployment(this, 'WebsiteFiles', {
+      destinationBucket: bucket,
+      sources: [Source.asset('./website')],
+      contentType: 'text/html',
+      retainOnDelete: false,
+    });
+
+    // create a cloudFront function from the request-function.js file
+    const requestFunction = new Function(this, 'RequestFunction', {
+      functionName: 'RequestFunction',
+      runtime: FunctionRuntime.JS_2_0,
+      code: FunctionCode.fromFile({
+        filePath: './cloudfront-functions/request-function.js'
+      })
+    });
+
+    // create a cloudFront function from the response-function.js file
+    const responseFunction = new Function(this, 'ResponseFunction', {
+      functionName: 'ResponseFunction',
+      runtime: FunctionRuntime.JS_2_0,
+      code: FunctionCode.fromFile({
+        filePath: './cloudfront-functions/response-function.js'
+      })
+    });
+
+    const cloudfrontOAI = new OriginAccessIdentity(this, 'cloudfront-OAI', {
```

Comment:
Can we use an OAC instead since that is best practice?

Reply:
I agree that OAC would be the best option here, but according to [this thread](https://github.com/aws/aws-cdk/issues/21771) we don't fully support it, as the `CloudFront Distribution` constructor always set a OAI by default when pointing to a S3 bucket as origin.
There are some work-around solutions to use OAC and remove the OAI authorization from the bucket policy, but the solution is not that "pretty", like [this one](https://github.com/aws/aws-cdk/issues/21771#issuecomment-1281190832).
What is your opinion on that? Should we implement this work-around solution or going for the not-the-best one?

Reply:
I guess we leave it as OAI, that's too bad CDK doesn't support OAC by default.